Added support for `TransmitFile` (on Windows) or `sendfile`.

jocelyn opened the pull request.

jocelyn committed:
Added support for `TransmitFile` (on Windows) or `sendfile`.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -29,12 +29,15 @@ indexing
 #define FD_SETSIZE 256
 #include <winsock2.h>
 #include <ws2tcpip.h>
+#include <Mswsock.h>
```
are we sure this file exists with VS or MinGW?

jocelyn replied:
On my machine, I have msc_vc140, and mingw coming with EiffelStudio.
And when ISE_C_COMPILER is set to either, msc_vc140 or mingw, in both cases library/net/Clib/ is compiled without error.

jocelyn replied:
> The Mswsock.h header file contains definitions for Microsoft-specific extensions to the Windows Sockets 2 (TransmitFile, AcceptEx, and ConnectEx, for example). The Mswsock.h header file is not normally needed unless these Microsoft-specific extensions are used by the application.

jocelyn replied:
Note that if I remove the include of `MSWSock.h`, and change the code to be

```
EIF_INTEGER c_sendfile_noexception(EIF_INTEGER out_fd, FILE* f, EIF_INTEGER length)
    /* transmission of file content from file `f` of size length through socket out_fd.
     * if `length` is 0, then send the whole file.
     *  NO exception is raised, and eventual error is return as result!
     */
{
#ifdef EIF_WINDOWS
    HMODULE dllHandle;
    BOOL (PASCAL * _TransmitFile)(SOCKET, HANDLE, DWORD, DWORD, LPOVERLAPPED, void*, DWORD) = NULL;

    dllHandle = LoadLibrary("Mswsock.dll");
    if (dllHandle != NULL) {
        _TransmitFile = (BOOL (PASCAL *) (SOCKET, HANDLE, DWORD, DWORD, LPOVERLAPPED, void*, DWORD))GetProcAddress(dllHandle, "TransmitFile");
    }
    if (_TransmitFile==NULL) {
        return (EIF_INTEGER) c_unoptimized_sendfile_noexception(out_fd, f, length);
    } else {
        BOOL retval;
        HANDLE hFile = (HANDLE)_get_osfhandle(fileno(f));
        OVERLAPPED ovlp;
        memset(&ovlp, 0, sizeof(OVERLAPPED));
        retval = _TransmitFile((EIF_SOCKET_TYPE) out_fd, (HANDLE) hFile, (DWORD) length, 
                    (DWORD) 0, (LPOVERLAPPED) &ovlp, (void*) NULL,(DWORD) 0x20 /*TF_USE_KERNEL_APC*/);
        if (retval) {
                /* Fix me, find a way to get the number of bytes sent */
            return (EIF_INTEGER) ovlp.InternalHigh;
        } else {
            return (EIF_INTEGER) -1;
        }
        FreeLibrary(dllHandle);
    }
```

i.e change `LPTRANSMIT_FILE_BUFFERS` with `void*`
and `TF_USE_KERNEL_APC` with 0x20

This is compiling and executing fine.
But this sounds more a hack than a clean solution.

I will check which version of VS include MSWSock.h

kwaxer replied:
Isn't it too much if a single send over network is accompanied with DLL load/unload? I'd better keep the function address somewhere and do not unload the DLL unless the Eiffel object to do the transmissions is released. Then the DLL can be unloaded.

kwaxer replied:
When return value is FALSE, the GetLastError should be checked for WSA_IO_PENDING or ERROR_IO_PENDING that would mean the operation has been initialiated, but not yet completed. And if the Eiffel library should tell if the operation has been completed, the C code should check that as well (according to the documentation, TransmitFile puts one object to the signaled state on completion).

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -743,7 +746,7 @@ EIF_INTEGER c_peer_name(EIF_INTEGER s, EIF_POINTER addr, EIF_INTEGER length)
 
 EIF_INTEGER c_send_stream_to_noexception(EIF_INTEGER fd, EIF_POINTER stream_pointer, EIF_INTEGER length, EIF_INTEGER flags, EIF_POINTER addr_pointer, EIF_INTEGER sizeofaddr)
-	/*  Transmission of string s of size size trought socket fd
+	/*  Transmission of string s of size size throught socket fd
```
let's fix this correctly, through

jocelyn replied:
oups

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -775,11 +778,108 @@ EIF_INTEGER c_put_stream_noexception(EIF_INTEGER fd, EIF_POINTER stream_pointer,
 	eif_net_check(c_put_stream_noexception(fd, stream_pointer, length));
 }
 
+EIF_INTEGER c_unoptimized_sendfile_noexception(EIF_INTEGER out_fd, FILE* f, EIF_INTEGER length)
```
Is there a reason for having both unoptimized and optimized version? Can we just have one version that knows what to take?

manuseiffel replied:
We should reject a length of 0 by default. If the file needs to be calculated, we should do that in the Eiffel code. This will avoid the need of refering `eif_file.h' which is runtime specific for EiffelBase and should not be used anywhere else.

jocelyn replied:
Do you mean you would prefer the two versions unoptimized and optimzed to be merged into a bigger one? We could do that, but I think it is cleaner to have 2 smaller functions that can be used by themselves, rather than a bigger one.

---

About `eif_file.h`
I thought the eif_...h file where the "public" part of the runtime.
And for now it is used for `eif_file_fd(f)`  and less critical `eif_file_size(f)`

manuseiffel replied:
I did not see the unoptimized one being used on my first review, so I thought the choice was made at the Eiffel level. I would call it `c_sendfile` and `c_sendfile_fallback`. No need for the `noexception` suffix because this is a a new API and we cannot add that suffix each time.

manuseiffel reviewed (commented):
Could you provide the Eiffel counterpart too?

jocelyn commented:
The Eiffel counterpart:

```
    put_file_content_noexception (a_file: FILE; a_offset: INTEGER; a_byte_count: INTEGER)
            -- Send `a_count' bytes from the content of file `a_file' starting at offset `a_offset'.
            -- If a_byte_count = 0, send the whole file.
        require
            byte_count_non_negative: a_byte_count >= 0
            extendible: extendible
        local
            l_close_needed: BOOLEAN
            l_bytes_sent: INTEGER
            cs: C_STRING
        do
            if a_file.exists and then a_file.is_access_readable then
                if a_file.is_open_read then
                    l_close_needed := False
                else
                    l_close_needed := True
                    a_file.open_read
                end
                if a_offset > 0 then
                    a_file.move (a_offset)
                end
                create cs.make (a_file.path.name)
                l_bytes_sent := c_sendfile_noexception (fd, a_file.file_pointer, a_byte_count)
                if l_bytes_sent < 0 then
                    report_error ("Network error: sendfile")
                end
                set_bytes_sent (l_bytes_sent)
                if l_close_needed then
                    a_file.close
                end
            end
        ensure
            bytes_sent_updated: (a_byte_count > 0 and not was_error) implies (0 <= bytes_sent and bytes_sent <= a_byte_count)
        end
```

And 

```
    c_sendfile_noexception (a_fd: INTEGER; fp: POINTER; len: INTEGER): INTEGER
            -- External routine to write file content from `fp` of
            -- length `len' to socket `a_fd'
            --      EIF_INTEGER c_sendfile_noexception(EIF_INTEGER out_fd, EIF_POINTER *fp, EIF_INTEGER length).
        external
            "C blocking"
        end
```

kwaxer commented:
-- Send `a_count' bytes...

Should it be "Send `a_byte_count' bytes ..."?

Also, using a special value 0 to indicate that a whole file is going to be send is not intuitive. Moreover, it is not the whole file if offset is not 0. I'd go for two procedures: a specific one with offset and amount, and a simpler one, just with file.

BTW, I'm not sure current implementation will work if a passed file is open for writing only. I believe it should be either closed or open for reading.

jocelyn commented:
> Also, using a special value 0 to indicate that a whole file is going to be send is not intuitive. Moreover, it is not the whole file if offset is not 0. I'd go for two procedures: a specific one with offset and amount, and a simpler one, just with file.

I am ok to remove the case "0" -> whole file. This was provided only by TransmitFile, but not "sendfile".
So let's remove it as it is not intuitive.
Otherwise, we can keep  `offset` and `amount`, for now.

> BTW, I'm not sure current implementation will work if a passed file is open for writing only. I believe it should be either closed or open for reading.
> Indeed, I will add this as precondition.

jocelyn commented:
There is a major open question for TransmitFile, indeed this is asynchronous operation.
And it seems to close the potential persistent connection in the case of EiffelWeb standalone server.

I don't know yet, how this should be added to EiffelNet, and then EiffelWeb.
For now, only the C function are added to network.c  
Should I also add the pending Eiffel code to EiffelNet ?

jocelyn commented:
Maybe I should only add this sendfile/transmitFile function to EiffelWeb for now.

jocelyn committed:
Updated c_sendfile_noexception and c_unoptimized_sendfile_noexception function.
Requires the input file `f` to be openread, and `length` > 0 (no more specific case when length is 0).
(note: not yet tested on non Windows platform).

jocelyn committed:
Fixed C compilation on non Windows platforms.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -775,11 +779,162 @@ EIF_INTEGER c_put_stream_noexception(EIF_INTEGER fd, EIF_POINTER stream_pointer,
+			return (EIF_INTEGER) -1;
+		}
+	} else {
+		static BOOL _TransmitFile_searched = FALSE;
```
This is not thread safe but why do you need to do a recursion once you calculated the address of TransmitFile?

manuseiffel replied:
Was reading the doc and you need to create an EVent in the OVERLAPPED structure and if you havE WSA_IO_PENDING, you just wait until the event is signaled. That way you are sure it is synchronous.

jocelyn committed:
Merge branch 'master' into net_sendfile

jocelyn committed:
Update EiffelNet Clib sendfile support.

jocelyn committed:
Include stdio on all platforms.

jocelyn committed:
Include sendfile usage in EiffelWeb code.

jocelyn committed:
Support offset parameter.

jocelyn committed:
cosmetic

jocelyn committed:
Use directly the TransmitFile .. without using LoadLibrary.

jocelyn commented:
Committed see https://svn.eiffel.com/websvn/revision.php?repname=eiffelstudio&rev=99467
